lib/index.ts

```diff
@@ -57,6 +57,141 @@ export async function inspect(
   return { plugin, package: pkg };
 }
 
+interface UpgradeRemediation {
+  upgradeTo: string;
+  // Other fields are of no interest
+}
+
+interface DependencyUpdates {
+  [from: string]: UpgradeRemediation;
+}
+
+interface ManifestFiles {
+  // Typically these are requirements.txt, constraints.txt and Pipfile;
+  // the plugin supports paths with subdirectories
+  [name: string]: string; // name-to-content
+}
+
+// Correction for the type; should be fixed in snyk-cli-interface
+export interface DepTreeDep {
+  name?: string; // shouldn't, but might be missing
+  version?: string; // shouldn't, but might be missing
+  dependencies?: {
+    [depName: string]: DepTreeDep;
+  };
+  labels?: {
+    [key: string]: string;
+
+    // Known keys:
+    // pruned: identical subtree already presents in the parent node.
+    //         See --prune-repeated-subdependencies flag.
+  };
+}
+
+// Applies upgrades to direct and indirect dependencies
+export async function applyRemediationToManifests(
+  root: string,
+  manifests: ManifestFiles,
+  upgrades: DependencyUpdates,
+  options: Options
+) {
+  const manifestNames = Object.keys(manifests);
+  const targetFile = manifestNames.find(
+    (fn) => path.basename(fn) === 'requirements.txt'
+  );
+  if (
+    !targetFile ||
+    !manifestNames.every(
+      (fn) =>
+        path.basename(fn) === 'requirements.txt' ||
+        path.basename(fn) === 'constraints.txt'
+    )
+  ) {
+    throw new Error(
+      'Remediation only supported for requirements.txt and constraints.txt files'
+    );
+  }
+
+  const provOptions = { ...options };
+  provOptions.args = provOptions.args || [];
+  provOptions.args.push('--only-provenance');
+
+  const topLevelDeps = (await inspect(root, targetFile, provOptions)).package;
+  applyUpgrades(manifests, upgrades, topLevelDeps);
+
+  return manifests;
+}
+
+function applyUpgrades(
+  manifests: ManifestFiles,
+  upgrades: DependencyUpdates,
+  topLevelDeps: DepTree
+) {
+  const requirementsFileName = Object.keys(manifests).find(
+    (fn) => path.basename(fn) === 'requirements.txt'
+  ) as string;
+  const constraintsFileName = Object.keys(manifests).find(
+    (fn) => path.basename(fn) === 'constraints.txt'
+  );
+
+  // Updates to requirements.txt
+  const patch: { [zeroBasedIndex: number]: string | false } = {}; // false means remove the line
+  const append: string[] = [];
+
+  for (const upgradeFrom of Object.keys(upgrades)) {
```

Comment:
this is not very readable with so many if else, can it be simplified with functions?

Reply:
I don't think functions would help here. The overhead on the repetitions and passing around parameters and having to scroll the code will not improve the readability.